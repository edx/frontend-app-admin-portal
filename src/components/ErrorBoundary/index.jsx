import React from 'react';
import PropTypes from 'prop-types';

import NewRelicService from '../../data/services/NewRelicService';

class ErrorBoundary extends React.Component {
  componentDidCatch(error) {
    NewRelicService.logError(error);
  }

  render() {
    return this.props.children;
  }
}

export const withErrorBoundary = WrappedComponent => props => (
  <ErrorBoundary>
    <WrappedComponent {...props} />
  </ErrorBoundary>
);

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ErrorBoundary;
