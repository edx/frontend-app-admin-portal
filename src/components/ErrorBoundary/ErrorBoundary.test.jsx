import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import NewRelicService from '../../data/services/NewRelicService';
import ErrorBoundary from './index';

global.newrelic = {
  addPageAction: jest.fn(),
  noticeError: jest.fn(),
};

function Welcome() {
  return <h1>Hi!</h1>;
}

describe('<ErrorBoundary />', () => {
  it('renders childern correctly when no error', () => {
    const tree = renderer
      .create((
        <ErrorBoundary>
          <Welcome name="MA" />
        </ErrorBoundary>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('logs the error', () => {
    const welcomeError = new Error('Catch me if you can!');
    NewRelicService.logError = jest.fn();

    const wrapper = shallow((
      <ErrorBoundary>
        <Welcome />
      </ErrorBoundary>
    ));
    wrapper.find(Welcome).simulateError(welcomeError);
    expect(NewRelicService.logError).toHaveBeenCalledWith(new Error('Catch me if you can!'));
  });
});
