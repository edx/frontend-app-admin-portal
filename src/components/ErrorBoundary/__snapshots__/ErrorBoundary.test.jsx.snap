// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> renders childern correctly when no error 1`] = `
<h1>
  Hi!
</h1>
`;
